projects/wallet-lib-client-for-event-queue/0001-event-queue-connection-design.md

```diff
@@ -0,0 +1,199 @@
+
+The `wallet-update` event will work with the schema:
+
+```ts
+interface WalletUpdateEvent {
+	type: 'wallet:address_history',
+	history: IHistoryTx,
+}
+
+// Where IHistoryTx is defined as:
+
+interface IHistoryTx {
+  tx_id: string;
+  signalBits: number;
+  version: number;
+  weight: number;
+  timestamp: number;
+  is_voided: boolean;
+  nonce: number,
+  inputs: IHistoryInput[];
+  outputs: IHistoryOutput[];
+  parents: string[];
+  token_name?: string;
+  token_symbol?: string;
+  tokens: string[];
+  height?: number;
+}
+
+export interface IHistoryInput {
+  value: number;
+  token_data: number;
+  script: string;
+  decoded: IHistoryOutputDecoded;
+  token: string;
+  tx_id: string;
+  index: number;
+}
+
+export interface IHistoryOutputDecoded {
+  type?: string;
+  address?: string;
+  timelock?: number | null;
+  data?: string;
+}
+
+export interface IHistoryOutput {
+  value: number;
+  token_data: number;
+  script: string;
+  decoded: IHistoryOutputDecoded;
+  token: string;
+  spent_by: string | null;
+}
+```
+
+The transaction data from the event queue is in a different format as described below:
+
+```ts
+interface EventQueueTxData {
+  hash: string;
+  version: number;
+  weight: number;
+  timestamp: number;
+  nonce?: number;
+  inputs: EventQueueTxInput[];
+  outputs: EventQueueTxOutput[];
+  parents: string[];
+  token_name?: string;
+  token_symbol?: string;
+  tokens: string[];
+  metadata: EventQueueTxMetadata;
+  aux_pow?: string;
+}
+
+interface EventQueueTxMetadata {
+  hash: string;
+  spent_outputs: EventQueueSpentOutput[];
+  conflict_with: string[];
+  voided_by: string[];
+  received_by: string[];
+  children: string[];
+  twins: string[];
+  accumulated_weight: number;
+  score: number;
+  first_block?: string;
+  height: number;
+  validation: string;
+}
+
+interface EventQueueTxInput {
+  tx_id: string;
+  index: number;
+  data: string;
+}
+
+interface EventQueueTxOutput {
+  value: number;
+  script: string;
+  token_data: number;
+}
+
+interface EventQueueSpentOutput {
+  index: number;
+  tx_ids: string[];
+}
+```
+
+### Data conversion process
+
+To keep compatibility with the current `Connection` class we need to convert the data with the following process:
+
+1. Convert `hash` to `tx_id`
+2. Assert `nonce` is valid
+3. Assign `signalBits` from `version`
+4. Assign `is_voided` from metadata's `voided_by`
+5. Assign `height` from metadata's height
+6. Convert outputs
+7. Convert inputs
+8. remove `metadata` and `aux_pow` fields
+
+Process to convert outputs:
+
+1. Derive `token` from `token_data` and `tx.tokens`
+2. Derive `spent_by` from the output index and `tx.metadata.spent_outputs`
+3. Derive `decoded` from the script
+
+Process to convert inputs:
+
+1. Try to find the transaction with the input's `tx_id` in storage
+	1. If not found we must fetch the tx from the fullnode api
```

Comment:
If `userA` spends inputs of `tx1` to send `tx2` to `userB`, we know `tx2` will be on both wallets, but `tx1` will only be on `userA`'s wallet.
The pubsub and transaction apis of the fullnode include data on all inputs since it has all transactions, but the event queue does not hidrate the transaction with more relevant input data which means we need to do this ourselves.

Basically, we need to fetch the data to do the same as these 2.

### Pubsub

When [publishing a tx](https://github.com/HathorNetwork/hathor-core/blob/master/hathor/indexes/address_index.py#L72) it uses the `to_json_extended` method which [hidrates the inputs with the spent output data](https://github.com/HathorNetwork/hathor-core/blob/master/hathor/transaction/base_transaction.py#L811-L818)

### Transaction API

It also uses `to_json_extended` for [list requests](https://github.com/HathorNetwork/hathor-core/blob/master/hathor/transaction/resources/transaction.py#L277) and it does a similar [hidration](https://github.com/HathorNetwork/hathor-core/blob/master/hathor/transaction/resources/transaction.py#L86-L91) with some extra data for single tx requests.